src/lib/crypto/kimchi_backend/gadgets/ec_group.ml

```diff
@@ -0,0 +1,3965 @@
+    ~label:"Ec_group.check_ia_neg_init_y"
+  @@ Affine.y neg_init_acc ;
+
+  (* Bounds 1: Input is public constant
+   *           Result is part of input (checked externally)
+   *)
+
+  (* C3: Copy computed_neg_init_acc to  ia.neg_acc *)
+  Affine.assert_equal (module Circuit) neg_init_acc ia.neg_acc ;
+
+  (* P is on curve <=> -P is on curve, thus we do not need to check
+   *  ai.neg_acc is on curve *)
+  ()
+
+(* Gadget for elliptic curve group scalar multiplication over foreign field
+ *
+ *   Given input point P and scalar field element s, computes and constrains that
+ *     Q = s0 * P + ... + sz * 2^z * P
+ *
+ *   where s0, s1, ..., sz is the binary expansion of s, (+) is group addition
+ *   and the terms P, 2 * P, ... 2^z * P are obtained with group doubling.
+ *
+ *   Inputs:
+ *      external_checks       := Context to track required external checks
+ *      curve                 := Elliptic curve parameters
+ *      scalar                := Boolean array of scalar bits
+ *      point                 := Affine point to scale
+ *
+ *   Preconditions and limitations:
+ *      P is not O (the point at infinity)
+ *      P's coordinates are bounds checked
+ *      P is on the curve
+ *      s is not zero
+ *      z > 0
+ *      ia point is randomly selected and constrained to be on the curve
+ *      ia negated point computation is constrained
+ *      ia coordinates are bounds checked
+ *
+ *   External checks: (per crumb, not counting inputs and output)
+ *     Bound checks:         42 (+1 when a != 0)
+ *     Multi-range-checks:   17
+ *     Compact-range-checks: 17
+ *     Total range-checks:   76
+ *
+ *   Rows: (per crumb, not counting inputs/outputs and constants)
+ *     Scalar multiplication:  ~84 (+2 when a != 0)
+ *     Bound additions:         84
+ *     Multi-range-checks:     308
+ *     Total:                  476
+ *
+ *   Constants:
+ *     Curve constants:        10 (for 256-bit curve; one-time cost per circuit)
+ *     Pre-computing doubles: 767 (for 256-bit curve; one-time cost per circuit)
+ *)
+let scalar_mul (type f) (module Circuit : Snark_intf.Run with type field = f)
+    (external_checks : f Foreign_field.External_checks.t)
+    (curve : f Curve_params.InCircuit.t) ?(doubles : f Affine.t array option)
+    (scalar : Circuit.Boolean.var list) (point : f Affine.t) : f Affine.t =
+  (* Double-and-add algorithm
+   *   Only used for signature verification, so simple algorithm suffices.
+   *
+   *     A = O; B = P
+   *     for i in 0..z
+   *         if si == 1
+   *             A = group_add(A, B)
+   *         B = group_double(B)
+   *     return A
+   *
+   *   Optimization:
+   *
+   *     To avoid expensive in-circuit conditional checks for point at infinity,
+   *     we employ a randomized strategy that avoids adding the identity element
+   *     or the same point to itself.  The strategy works as follows.
+   *
+   *     Since the prover knows the the points that it wewill add and double during
```

Comment:
Since the prover knows the points that it will add and double during

Reply:
nice spot!